src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -190,36 +211,45 @@ public virtual async Task<CircuitHost> ConnectAsync(string circuitId, IClientPro
 
         protected virtual (CircuitHost circuitHost, bool previouslyConnected) ConnectCore(string circuitId, IClientProxy clientProxy, string connectionId)
         {
-            if (ConnectedCircuits.TryGetValue(circuitId, out var circuitHost))
+            if (ConnectedCircuits.TryGetValue(circuitId, out var connectedCircuitHost))
             {
+                Log.ConnectingToActiveCircuit(_logger, circuitId, connectionId);
+
                 // The host is still active i.e. the server hasn't detected the client disconnect.
                 // However the client reconnected establishing a new connection.
-                circuitHost.Client.Transfer(clientProxy, connectionId);
-                return (circuitHost, true);
+                connectedCircuitHost.Client.Transfer(clientProxy, connectionId);
+                return (connectedCircuitHost, true);
             }
 
-            if (DisconnectedCircuits.TryGetValue(circuitId, out circuitHost))
+            if (DisconnectedCircuits.TryGetValue(circuitId, out DisconnectedCircuitEntry disconnectedEntry))
             {
+                Log.ConnectingToDisconnectedCircuit(_logger, circuitId, connectionId);
+
                 // The host was in disconnected state. Transfer it to ConnectedCircuits so that it's no longer considered disconnected.
-                DisconnectedCircuits.Remove(circuitId);
-                ConnectedCircuits.TryAdd(circuitId, circuitHost);
+                // First discard the CancellationTokenSource so that the cache entry does not expire.
+                DisposeTokenSource(disconnectedEntry);
 
-                circuitHost.Client.Transfer(clientProxy, connectionId);
+                DisconnectedCircuits.Remove(circuitId);
+                ConnectedCircuits.TryAdd(circuitId, disconnectedEntry.CircuitHost);
 
-                return (circuitHost, false);
+                disconnectedEntry.CircuitHost.Client.Transfer(clientProxy, connectionId);
+                return (disconnectedEntry.CircuitHost, false);
             }
 
             return default;
         }
 
-        private void OnEntryEvicted(object key, object value, EvictionReason reason, object state)
+        protected virtual void OnEntryEvicted(object key, object value, EvictionReason reason, object state)
         {
             switch (reason)
             {
                 case EvictionReason.Expired:
```

Comment:
Is the `Expired` case still possible?

Reply:
We shouldn't. I'll remove it so we know if it's being weird